### PR TITLE
Initial endpoint DDoS cooperation text

### LINKED ISSUE #240: Endpoint cooperation for DoS mitigation
[Section 4.4](https://quicwg.org/ops-drafts/draft-ietf-quic-manageability.html#section-4.4) talks at length about detecting and mitigating against (D)DoS attacks, but this is written with the assumption of an uncooperative endpoint (likely to be, but unstated, a server).  All except the Retry text where the design specifically forces interventions to be cooperative.

It is entirely reasonable for DoS classification to be done with full knowledge of valid connection IDs or at least valid connection ID lengths.  If you allow for cooperation, the text might then explain how this might allow established flows to be separated from spoofed flows and new connection attempts.

### PR DESCRIPTION
**Pull request by britram**
fixes #240. thanks @gorryfair for the text.

### CODE REVIEW

File: draft-ietf-quic-manageability.md [hunk 1]
```diff
@@ -779,7 +779,7 @@ not the network RTT.
 Since the spin bit logic at each endpoint considers only samples from packets
 that advance the largest packet number, signal generation itself is
 resistant to reordering. However, reordering can cause problems at an observer
-by causing spurious edge detection and therefore inaccurate (i.e., lower) RTT
+by causing spurious edge \ and therefore inaccurate (i.e., lower) RTT
```
Comment (mirjak) [inline draft-ietf-quic-manageability.md:782]:
```suggestion
by causing spurious edge detection and therefore inaccurate (i.e., lower) RTT
```

File: draft-ietf-quic-manageability.md [hunk 2]
```diff
@@ -945,37 +945,59 @@ severe packet loss.
...
-attack. If the connection migration is not visible to the network that performs
-the DDoS detection, an active, migrated QUIC connection may be blocked by such a
-system under attack. As soon as the connection blocking is detected by the
-client, the client may rely on the fast resumption mechanism provided by
-QUIC. When clients migrate to a new path, they should be prepared for the
-migration to fail and attempt to reconnect quickly.
-
-TCP syncookies {{?RFC4937}} are a well-established method of mitigating some
+ID. In this case, the connection ID can not be used to distinguish valid, active
+traffic from new attacl traffic.
```
Comment (nibanks) [inline draft-ietf-quic-manageability.md:979]:
```suggestion
traffic from new attack traffic.
```

File: draft-ietf-quic-manageability.md [hunk 3]
```diff
@@ -945,37 +945,59 @@ severe packet loss.
...
+
+Alternatively, when an encrypted transport is used, it is also possible for
+endpoints to directly support these security functions.
```
Comment (nibanks) [inline draft-ietf-quic-manageability.md:982]: I don't quite understand this sentence. For one, why "when an encrypted transport is used"? This document is purely about an encrypted transport, right? And then "these security functions" is also unclear to me.

File: draft-ietf-quic-manageability.md [hunk 4]
```diff
@@ -945,37 +945,59 @@ severe packet loss.
...
+Endpoints can cooperate with an on-path device directly by e.g. sharing
+information about connection IDs.
+
+Another potential method could use an
+on-path network device that relies on pattern inferences in the traffic and
+heuristics or machine learning instead of processing observed header
+information.
+
+However, it is questionable whether connection migrations must be supported
```
Comment (nibanks) [inline draft-ietf-quic-manageability.md:991]: Is this referring to just explicit connection migration or do you include NAT rebinding as well? Either way, I think it should be explicitly called out. I think, ideally, NAT rebinding should continue to work, but I tend to agree full, explicit connection migration likely isn't completely necessary while under attack.

File: draft-ietf-quic-manageability.md (hunk 3 above)
Comment (gorryfair) [reply draft-ietf-quic-manageability.md:982]: On the first: I think this is true for any encrypted transport, but yes it does apply to QUIC.  
On the functions: I thought this was: DOS;DDOS;Malware detection; NIDS; etc

File: draft-ietf-quic-manageability.md [hunk 5]
```diff
@@ -945,37 +945,59 @@ severe packet loss.
...
+security functions. For example, Denial of Service (DOS) and Distributed DOS
+(DDOS) attacks against the infrastructure or against an endpoint can be
+detected and mitigated by characterising anomalous traffic.
+Other uses include support for security audits (e.g., verifying the
+compliance with ciphersuites); client and application fingerprinting for
+inventory; and to provide alerts for network intrusion detection and other
+next generation firewall functions.
+
 Current practices in detection and mitigation of Distributed Denial of Service
 (DDoS) attacks generally involve classification of incoming traffic (as
```
Comment (britram) [inline draft-ietf-quic-manageability.md:958]:
```suggestion
Current practices in detection and mitigation of DDoS
attacks generally involve classification of incoming traffic (as
```

File: draft-ietf-quic-manageability.md [hunk 6]
```diff
@@ -945,37 +945,59 @@ severe packet loss.
...
+Efficient classification of this DDoS traffic in the mitigation environment
+is key to the success of this approach. Limited first-packet garbage detection
+as in {{sec-garbage}} and stateful tracking of QUIC traffic as in
+{{sec-stateful}} above may be useful during classification.
 
 Note that the use of a connection ID to support connection migration renders
 5-tuple based filtering insufficient and requires more state to be maintained by
-DDoS defense systems. For the common case of NAT rebinding, DDoS defense systems
+DDoS defense systems. For the common case of NAT rebinding, where the client's
+address changes without the client's intent or knowedge, DDoS defense systems
```
Comment (britram) [inline draft-ietf-quic-manageability.md:974]:
```suggestion
address changes without the client's intent or knowledge, DDoS defense systems
```

File: draft-ietf-quic-manageability.md [hunk 7]
```diff
@@ -945,37 +945,59 @@ severe packet loss.
...
-system under attack. As soon as the connection blocking is detected by the
-client, the client may rely on the fast resumption mechanism provided by
-QUIC. When clients migrate to a new path, they should be prepared for the
-migration to fail and attempt to reconnect quickly.
-
-TCP syncookies {{?RFC4937}} are a well-established method of mitigating some
+ID. In this case, the connection ID can not be used to distinguish valid, active
+traffic from new attack traffic.
+
+Alternatively, when an encrypted transport is used, it is also possible for
```
Comment (britram) [inline draft-ietf-quic-manageability.md:981]:
```suggestion
It is also possible for
```

File: draft-ietf-quic-manageability.md [hunk 8]
```diff
@@ -945,37 +945,59 @@ severe packet loss.
...
+endpoints to directly support these security functions.
```
Comment (britram) [inline draft-ietf-quic-manageability.md:982]:
```suggestion
endpoints to directly support security functions.
```

File: draft-ietf-quic-manageability.md [hunk 9]
```diff
@@ -945,37 +945,59 @@ severe packet loss.
...
+Endpoints can cooperate with an on-path device directly by e.g. sharing
+information about connection IDs.
```
Comment (britram) [inline draft-ietf-quic-manageability.md:984]:
```suggestion
Endpoints can also cooperate with an on-path device directly by e.g.
sharing information about connection IDs.
```

File: draft-ietf-quic-manageability.md [hunk 10]
```diff
@@ -945,37 +945,59 @@ severe packet loss.
...
+
+Another potential method could use an
+on-path network device that relies on pattern inferences in the traffic and
+heuristics or machine learning instead of processing observed header
+information.
+
+However, it is questionable whether connection migrations must be supported
+during a DDoS attack. If the connection migration is not visible to the network
```
Comment (mirjak) [inline draft-ietf-quic-manageability.md:992]:
```suggestion
during a DDoS attack. However, for unintentional mitigations, as usually the case
for NAT rebindings, the connection ID can easily be used to track connections.
If the connection migration is not visible to the network
```

File: draft-ietf-quic-manageability.md [hunk 11]
```diff
@@ -945,37 +945,63 @@ severe packet loss.
...
+information.
+
+However, it is questionable whether connection migrations must be supported
+during a DDoS attack. However, for unintentional mitigations, as is usually the
+case for NAT rebindings, the connection ID can be used to track
+connections (see {{sec-flow-association}}). If the connection
+migration is not visible to the network
+that performs the DDoS detection, an active, migrated QUIC connection may be
+blocked by such a system under attack. As soon as the connection blocking is
+detected by the client, the client may rely on the fast resumption mechanism
```
Comment (martinthomson) [inline draft-ietf-quic-manageability.md:999]: Resumption might be fast, but still costs a lot more to handle than keeping a connection alive.  This won't produce a good outcome in the case of DoS.  I don't think that this is good advice at all.

**Review by martinthomson — COMMENTED** [review]
I don't like where this was taken.  As @mirjak said, it is not responsive to my comments about cooperation.  It still firmly assumes that DoS mitigation is being performed without cooperation from endpoints.  That might be unavoidable for good reasons, but we should be clearer about the consequences of that AND we should recommend that DoS mitigation be performed with cooperation from endpoints.

**Pull request merged**